### PR TITLE
HZN-1464: Provide core API and module structure for a new graph service

### PR DESCRIPTION
**Pull request by mvrueden**
This PR provides a new module `features/graph` which at some point will implement a new Graph Service with more capabilities and probably better performance than our current topology implementation.

This is the first of a series of PRs and is just the beginning. 
A lot of this will probably change during further development.
It also has a lot of TODOs which need to be figured out.
For better "control" this is merged into a `features/graph-service` feature branch.

This PR mostly focusses on:
 - module structure
 - basic concepts (e.g. `GraphContainerProvider` vs `GraphProvider`)
 - persistence (Hibernate for now)
 - Calculating changes of containers/graphs

It also implements the following providers:
 - Dummy provider, which just has static vertices
 - GraphML provider. Anything you push through rest to the current graphml endpoint will be exposed via the new service
 - BSM provider. Any BSM graph will be exposed as well

You can test this by invoking `graph:list` and then `graph:get --container <containerId> --namespace <graphNamespace>` from the Karaf console.

NOTE: The relevant features are not installed automatically, you have to install them manually. 
The following ones are available as of now:
 - `opennms-graphs` (API and for now Impl)
 - `opennms-graph-provider-dummy` (Dummy Provider)
 - `opennms-graph-provider-graphml` (GraphML Provider)
 - `opennms-graph-provider-bsm` (BSM Provider)

It does not implement:
 - Enrichment (e.g. Status calculation, Node or IpInterface enrichment)
 - Searching or Suggestion Providers
 - Persisting "detached vertices", e.g. if an edge points to a vertex from another graph (especially outside its container)
 - Persisting collections of the supported types
 - Lazy Loading of graphs
 - ReST Service
 - Replacement providers for our existing TopologyProviders, e.g. VmWare
 - Listening for updates

However, for some of the above not yet implemented features, the API is already in place.

Please feel free to comment, however changes may be applied at a later state.

### CODE REVIEW

File: container/features/src/main/resources/features.xml [hunk 1]
```diff
@@ -1329,4 +1329,43 @@
         <feature version="${opennmsApiVersion}">opennms-integration-api</feature>
         <bundle>mvn:org.opennms.features/org.opennms.features.api-layer/${project.version}</bundle>
     </feature>
+
+    <!-- TODO MVR may split in several features , e.g. API / IMPL, etc.-->
+    <!-- TODO MVR redefine me -->
+    <feature name="opennms-graphs" description="OpenNMS :: Features :: Graphs" version="${project.version}">
```
[the remark this answers is not in the document]
Comment (mvrueden) [reply container/features/src/main/resources/features.xml]: Yeah, that was the plan. I omitted it for now, as the name is not final.

File: features/graph/shell/src/main/java/org/opennms/netmgt/graph/shell/GraphGetCommand.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+@Command(scope = "graph", name = "get", description="Gets a graph identified by its namespace")
+@Service
+// TODO MVR expose via xml (graphml) or json
+public class GraphGetCommand implements Action {
+
+    @Reference
+    private GraphService graphService;
+
+    @Option(name="--container", description="The id of the container", required=true)
+    private String containerId;
```
[the remark this answers is not in the document]
Comment (mvrueden) [reply features/graph/shell/src/main/java/org/opennms/netmgt/graph/shell/GraphGetCommand.java]: Autocompletion makes sense.

File: features/graph/api/src/main/java/org/opennms/netmgt/graph/api/service/GraphContainerProvider.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+public interface GraphContainerProvider {
+    // TODO MVR we have to implement this somehow
+//    /**
+//     * The provider may need to inform about graph changes.
+//     * Whith this method the {@link GraphNotificationService} is passed to the provider.
+//     * @param notificationService
+//     */
+//    void setNotificationService(GraphNotificationService notificationService);
+
+    GraphContainer loadGraphContainer();
```
[the remark this answers is not in the document]
Comment (mvrueden) [reply features/graph/api/src/main/java/org/opennms/netmgt/graph/api/service/GraphContainerProvider.java]: It should not load the container, but implement the loading. Caching is handled in the GraphService later on. We may need to add support for custom caching strategies. For example in the case of BSM there is no need to evict the cache regularly. Only if the daemon reload event was fired, etc. But that is not fully shaped out here.

I am going to add some javadoc to make the responsibilities more clear.

File: container/features/src/main/resources/features.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (j-white) [reply container/features/src/main/resources/features.xml]: I just thought it would be useful to have it loaded by default when testing this branch.

**Comment by mvrueden** [comment]
I am going forward am will merge this, as this is WIP anyways.

**Pull request merged**